From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 24 Mar 2019 01:01:32 -0400
Subject: [PATCH] Only count Natural Spawned mobs towards natural spawn mob
 limit

This resolves the super common complaint about mobs not spawning.

This was ultimately a flaw in the vanilla count algorithim that allows
spawners and other misc mobs to count against the mob limit, which are
not bounded, and can prevent the entire world from spawning new.

I believe Bukkits changes around persistence may of actually made it
worse than vanilla.

This should fully solve all of the issues around it so that only natural
influences natural spawns.

diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index b9410e3affed4004548acdc9355ac14d18c4a2c8..90e8defde0ef84f3146c8664805d6f11b84c593d 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -458,6 +458,16 @@ public class PaperWorldConfig {
         maxAutoSaveChunksPerTick = getInt("max-auto-save-chunks-per-tick", 24);
     }
 
+    public boolean countAllMobsForSpawning = false;
+    private void countAllMobsForSpawning() {
+        countAllMobsForSpawning = getBoolean("count-all-mobs-for-spawning", false);
+        if (countAllMobsForSpawning) {
+            log("Counting all mobs for spawning. Mob farms may reduce natural spawns elsewhere in world.");
+        } else {
+            log("Using improved mob spawn limits (Only Natural Spawns impact spawn limits for more natural spawns)");
+        }
+    }
+
     public boolean antiXray;
     public EngineMode engineMode;
     public int maxChunkSectionIndex;
diff --git a/src/main/java/net/minecraft/world/spawner/WorldEntitySpawner.java b/src/main/java/net/minecraft/world/spawner/WorldEntitySpawner.java
index 84e95bdd7ba473d2c8c884f9e71861d33747900a..5f27b01ef8af05c714fbde2311f147cfc82ac04e 100644
--- a/src/main/java/net/minecraft/world/spawner/WorldEntitySpawner.java
+++ b/src/main/java/net/minecraft/world/spawner/WorldEntitySpawner.java
@@ -88,6 +88,13 @@ public final class WorldEntitySpawner {
             EntityClassification enumcreaturetype = entity.func_200600_R().func_220339_d();
 
             if (enumcreaturetype != EntityClassification.MISC) {
+                // Paper start - Only count natural spawns
+                if (!entity.field_70170_p.paperConfig.countAllMobsForSpawning &&
+                    !(entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL ||
+                        entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN)) {
+                    continue;
+                }
+                // Paper end
                 BlockPos blockposition = entity.func_233580_cy_();
                 long j = ChunkPos.func_77272_a(blockposition.func_177958_n() >> 4, blockposition.func_177952_p() >> 4);
 
